app/models/concerns/cached_counting.rb

```diff
@@ -3,75 +3,160 @@
 module CachedCounting
   extend ActiveSupport::Concern
 
-  EXPIRE_CACHE_AFTER = 4.days.to_i
-
-  LUA_INCR_AND_EXPIRE = DiscourseRedis::EvalHelper.new <<~LUA
-    local result = redis.call("INCR", KEYS[1])
-
-    if result == 1 then
-      redis.call("EXPIRE", KEYS[1], ARGV[1])
-    end
+  LUA_HGET_DEL = DiscourseRedis::EvalHelper.new <<~LUA
+    local result = redis.call("HGET", KEYS[1], KEYS[2])
+    redis.call("HDEL", KEYS[1], KEYS[2])
 
     return result
   LUA
 
-  included do
-    class << self
-      attr_accessor :autoflush, :autoflush_seconds, :last_flush
-    end
+  QUEUE = Queue.new
+  SLEEP_SECONDS = 1
+  FLUSH_DB_ITERATIONS = 60
+  MUTEX = Mutex.new
 
-    # auto flush if backlog is larger than this
-    self.autoflush = 2000
+  def self.disable
+    @enabled = false
+    if @thread && @thread.alive?
+      @thread.wakeup
+      @thread.join
+    end
+  end
 
-    # auto flush if older than this
-    self.autoflush_seconds = 5.minutes
+  def self.enabled?
+    @enabled != false
+  end
 
-    self.last_flush = Time.now.utc
+  def self.enable
+    @enabled = true
   end
 
-  class_methods do
-    def perform_increment!(key, opts = nil)
-      val = DiscourseRedis.ignore_readonly do
-        LUA_INCR_AND_EXPIRE.eval(
-          Discourse.redis.without_namespace,
-          [Discourse.redis.namespace_key(key)],
-          [EXPIRE_CACHE_AFTER]
-        ).to_i
-      end
+  def self.reset
+    clear_queue!
+    clear_flush_to_db_lock!
+  end
 
-      # readonly mode it is going to be nil, skip
-      return if val.nil?
+  def self.ensure_thread!
+    return if !enabled?
 
-      autoflush = (opts && opts[:autoflush]) || self.autoflush
-      if autoflush > 0 && val >= autoflush
-        write_cache!
-        return
+    MUTEX.synchronize do
+      if !@thread&.alive?
+        @thread = nil
       end
+      @thread ||= Thread.new { thread_loop }
+    end
+  end
 
-      if (Time.now.utc - last_flush).to_i > autoflush_seconds
-        write_cache!
+  def self.thread_loop
+    iterations = 0
+    while true
+      break if !enabled?
+
+      sleep SLEEP_SECONDS
```

Comment:
This means it will take up-to-1s every time we want to force a flush, or cleanly stop the thread (mostly a problem in tests, I guess).

Could we use an 'interruptible sleep' implementation instead? Like this:

https://github.com/discourse/discourse-prometheus/blob/d71565f7ee4d3fe5cef8c8831a20cec5e52a1367/lib/reporter/global.rb#L24-L43

Reply:
Oh, I just realised that `Thread#awake` takes care of this problem. TIL!